internal/controller/sparkapplication/controller.go

```diff
@@ -695,6 +724,146 @@ func (r *Reconciler) reconcileUnknownSparkApplication(ctx context.Context, req c
+			if err != nil {
+				return err
+			}
+			if old.Status.AppState.State != v1beta2.ApplicationStateSuspending {
+				return nil
+			}
+			app := old.DeepCopy()
+
+			r.recordSparkApplicationEvent(app)
+
+			if err := r.deleteSparkResources(ctx, app); err != nil {
+				logger.Error(err, "failed to delete spark resources", "name", app.Name, "namespace", app.Namespace)
+				return err
+			}
+
+			app.Status.AppState = v1beta2.ApplicationState{
+				State: v1beta2.ApplicationStateSuspended,
+			}
+			r.resetSparkApplicationStatus(app)
+
+			if err := r.updateSparkApplicationStatus(ctx, app); err != nil {
+				return err
+			}
+			return nil
+		},
+	)
+	if retryErr != nil {
+		logger.Error(retryErr, "Failed to reconcile SparkApplication", "name", key.Name, "namespace", key.Namespace)
+		return ctrl.Result{}, retryErr
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) reconcileSuspendedSparkApplication(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	key := req.NamespacedName
+	retryErr := retry.RetryOnConflict(
+		retry.DefaultRetry,
+		func() error {
+			old, err := r.getSparkApplication(ctx, key)
+			if err != nil {
+				return err
+			}
+			if old.Status.AppState.State != v1beta2.ApplicationStateSuspended {
+				return nil
+			}
+			app := old.DeepCopy()
+
+			if r.validateSparkResourceDeletion(ctx, app) {
+				logger.Info("Successfully deleted resources associated with SparkApplication", "name", app.Name, "namespace", app.Namespace, "state", app.Status.AppState.State)
+				r.resetSparkApplicationStatus(app)
+				r.recordSparkApplicationEvent(app)
+				if !ptr.Deref(app.Spec.Suspend, false) {
+					app.Status.AppState = v1beta2.ApplicationState{
+						State: v1beta2.ApplicationStateResuming,
+					}
+				}
+			} else {
+				err := fmt.Errorf("resources associated with SparkApplication still exist: %s/%s", app.Namespace, app.Name)
+				logger.Error(err, "Failed to confirm being deleted resources associated with SparkApplication", "name", app.Name, "namespace", app.Namespace, "state", app.Status.AppState.State)
+				return err
+			}
+			if err := r.updateSparkApplicationStatus(ctx, app); err != nil {
+				return err
+			}
+			return nil
+		},
+	)
+	if retryErr != nil {
+		logger.Error(retryErr, "Failed to reconcile SparkApplication", "name", key.Name, "namespace", key.Namespace)
+		return ctrl.Result{}, retryErr
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) reconcileResumingSparkApplication(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	key := req.NamespacedName
+	retryErr := retry.RetryOnConflict(
+		retry.DefaultRetry,
+		func() error {
+			old, err := r.getSparkApplication(ctx, key)
+			if err != nil {
+				return err
+			}
+			if old.Status.AppState.State != v1beta2.ApplicationStateResuming {
+				return nil
+			}
+			app := old.DeepCopy()
+
+			r.recordSparkApplicationEvent(app)
+
+			_ = r.submitSparkApplication(ctx, app)
```

Comment:
Don't you need to fail on error here and retry? 

If the handling of error is not necessary, it is probably worth adding a comment why, because it is far from obvious.